### PR TITLE
enable ddp as a plugin

### PR DESCRIPTION
**Pull request by williamFalcon**
It's annoying to set ddp for all accelerators when configuring a custom ddp. Instead, you can now pass in a plugin.

Bonus: Plugin docs 

cc @ananthsub

### CODE REVIEW

File: pytorch_lightning/accelerators/ddp_accelerator.py [hunk 1]
```diff
@@ -47,7 +46,7 @@
 class DDPAccelerator(Accelerator):
 
     def __init__(self, trainer, cluster_environment=None):
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/ddp_accelerator.py:48]: ddp_plugin is missing from the constructor here

File: pytorch_lightning/accelerators/ddp_accelerator.py [hunk 2]
```diff
@@ -287,9 +286,7 @@ def ddp_train(self, process_idx, model):
     def configure_ddp(
         self, model: LightningModule, device_ids: List[int]
     ) -> DistributedDataParallel:
-        model = LightningDistributedDataParallel(
-            model, device_ids=device_ids, find_unused_parameters=True
-        )
+        model = self.ddp_plugin.configure_ddp(model, device_ids)
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/ddp_accelerator.py:289]: do you need to check that `self.ddp_plugin` is not `None` ? it'd be nice to have the typehint for it as `Optional[DDPPlugin]`

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    Example::
+
+        class MyDDP(DDPPlugin):
+
+            def configure_ddp(self, model, device_ids):
+                model = MyDDPWrapper(model, device_ids)
+                return model
+
+        my_ddp = MyDDP()
+        trainer = Trainer(accelerator='ddp_x', plugins=[my_ddp])
```
Comment (ananthsub) [inline pytorch_lightning/plugins/ddp_plugin.py:19]: super nice!

**Review by ananthsub — COMMENTED** [review]
this is grerat! i assume sync batch norm would follow the same plugin pattern

**Comment by williamFalcon** [comment]
> this is great! i assume sync batch norm would follow the same plugin pattern

yup! 
and anything else that should be accelerator agnostic as well.

I think this is a general work in progress as we figure out what should be moved into plugins to increase flexibility.
We already have all the half precision backends as plugins, now ddp and synced batchnorm as well.

I assume in a few months we'll again condense whatever new things have come out and make this API a bit more future proof / general

File: pytorch_lightning/accelerators/ddp_accelerator.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/accelerators/ddp_accelerator.py:289]: the root accelerator ensures this should not be none.

but tests are failing for some reason where one is none.... trying to figure out why

Comment (williamFalcon) [reply pytorch_lightning/accelerators/ddp_accelerator.py:289]: found it... pickle state snuck in there haha

**Review by ananthsub — APPROVED** [review]

**Comment by SeanNaren** [comment]
This is awesome :) it might be worth adding something to the docs to clear up that custom DDP implementations forward functions need to handle more than just forward input tensor arguments in PL: https://github.com/PyTorchLightning/pytorch-lightning/blob/feature/817-fairscale/pytorch_lightning/overrides/data_parallel.py#L162

**Comment by williamFalcon** [comment]
> This is awesome :) it might be worth adding something to the docs to clear up that custom DDP implementations forward functions need to handle more than just forward input tensor arguments in PL: https://github.com/PyTorchLightning/pytorch-lightning/blob/feature/817-fairscale/pytorch_lightning/overrides/data_parallel.py#L162

i’ll add in a follow on PR

**Comment by Borda** [comment]
@edenlightning is it `1.1` or `1.0` feature?

File: pytorch_lightning/accelerators/accelerator.py [hunk 4]
```diff
@@ -36,11 +36,12 @@
 
 class Accelerator(object):
 
-    def __init__(self, trainer=None, cluster_environment=None):
+    def __init__(self, trainer=None, cluster_environment=None, ddp_plugin=None):
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:39]: revisiting this for https://github.com/PyTorchLightning/pytorch-lightning/issues/4349

Could we read `ddp_plugin` directly off `trainer.plugn_connector`? otherwise as we add more plugin types, this constructor will be huge
